### DISCUSSION TITLE
Allow spaces in enum values (server side regex doesn't allow spaces)

### DISCUSSION THREAD

**Discussion by websocket98765** [strapi/strapi#4780] [Legacy Topics]
**Describe the bug**

Given enums are strings, whether a space exists or not should be irrelevant (e.g. 'Hewlett Packard').

In the past (Strapi 3.0.0-beta.17.x & earlier), it was possible to specify enum values containing spaces as long as this was done in the `Foo.settings.json`, and not via the UI. In Strapi 3.0.0-beta.18.2, Strapi appears to now enforce this server-side via a regex, as I receive a 400 Bad Request when trying to save my Content Type in the UI and an error message in the browser's console log contains a number of these: `"contentType.attributes.foo.enum[3] must match the following regex: /^[_A-Za-z][_0-9A-Za-z]*$/"` (Presumably because some of my enum values contain a space.)

This may be expected behavior, but if so, I'd like to create a discussion that, from a user perspective, it would be preferrable to allow spaces in enum values (e.g. a dropdown of brand names like 'Hewlett Packard'), given they are merely pre-defined strings.

**Steps to reproduce the behavior**
1. Edit Foo.settings.json so an enum field has one or more values containing a space.
2. In Strapi's UI, go to Content Type editing section for Foo, and try to make a change to any item. It shouldn't matter what is changed. Click "save".
3. "An error occurred" will be shown. 
![error occurred](https://user-images.githubusercontent.com/41795874/71229000-ec8bf980-22db-11ea-94a7-23870c14a4c8.png)
4. See the browser's console log for more error text: ``"contentType.attributes.foo.enum[3] must match the following regex: /^[_A-Za-z][_0-9A-Za-z]*$/"``

**Expected behavior**

The server-side validation should allow spaces in enum values.

**System**
- Node.js version: v13
- NPM version: 6.13.4
- Strapi version: 3.0.0-beta.18.2
- Database: Sqlite
- Operating system: MacOS

**Comment by derrickmehaffy** [comment]
This is not an issue with Strapi directly but one that comes from GraphQL (even if you aren't using graphql, it's enforced for all users to allow for easy installation of the GQL plugin).

As Facebook designed GQL, the rules are enforced and this is not considered a bug. I would recommend you use camelCase for enumeration values.

For now going to mark this closed as it's not a bug and is intended.

**Comment by derrickmehaffy** [comment]
For reference:


> Name :: /[_A-Za-z][_0-9A-Za-z]*/
> GraphQL query documents are full of named things: operations, fields, arguments, directives, fragments, and variables. All names must follow the same grammatical form. Names in GraphQL are case‐sensitive. That is to say name, Name, and NAME all refer to different names. Underscores are significant, which means other_name and othername are two different names. Names in GraphQL are limited to this ASCII subset of possible characters to support interoperation with as many other systems as possible.


http://facebook.github.io/graphql/#sec-Names

**Comment by derrickmehaffy** [comment]
https://github.com/graphql/graphql-js/issues/936#issuecomment-349887156

**Comment by derrickmehaffy** [comment]
One more bit of clarification, this topic has been discussed at length:
https://github.com/strapi/strapi/issues/4250
https://github.com/strapi/strapi/issues/3921
https://github.com/strapi/strapi/issues/3340
https://github.com/strapi/strapi/issues/2494
https://github.com/strapi/strapi/issues/1698

**Comment by websocket98765** [comment]
The long list of user-reported enum issues is a good evidence to how this deviates from expected user behavior.

Two ways to improve this:
1. A potential solution would be for Strapi to basically tell GraphQL it's a string type, given it arrives in the database as a string anyway
2. Irrespective of the decision on the above, this related issue <https://github.com/strapi/strapi/issues/2494> is _UI related_ and should not be closed for the stated reason, at least, because it actually does _not_ run into GraphQL's enum limitations as stated. The OP is describing a different way to generate the UI element for enums, so that label and value can be different. This is useful, especially if enums have enforced current character limitations. The issue is _just about creating the UI element_ using the label for display. The label is UI related; it isn't sent to the DB and does not need to pass any GraphQL validation. This is actually a valuable idea and something @soupette should consider. (Note for @soupette: it's possible to support this without a breaking change by checking whether the enum definition contains objects or strings when building the UI element).

**Comment by derrickmehaffy** [comment]
You are more than welcome to submit a PR

**Comment by alexandrebodin** [comment]
Hi @websocket98765,

I agree that this is not user firendly at all.
There are multiple things we can consider arround the enum feature:

## 1
Using the values has labels and create normalized names that will work with graphQL or more restrictive DBs.

## 2
Allow for a label - value pair to be set. 

--- 
We could also consider adding multiple choice option

In any case we need to consider the tradeoffs and advantages of those solutions. 

Feel free to explain what would be the best solution for you.

We chose to be restrictive for now to prevent issues with graphql like you mentionned. But we will remove this constraint as soon as we can :)

**Comment by websocket98765** [comment]
@alexandrebodin Is there a reason storing as a string wouldn't work? That would be the best solution imo.

What we're really dealing with are user-defined string values--people use dropdowns to provide a limited &/or quick choice of options, to prevent misspellings, etc. (Enums don't really exist in JS. They're used in C++ to map a human-readable name to a 4-bit integer for better runtime performance. In JS, it’s simply a string. Telling GraphQL that's what it is would solve the issue cleanly.) IMO, the expected user experience from a dropdown would be to allow any string value (so it's possible to add a value like "Hewlett Packard").

The "label - value pair" idea is tangential and doesn't directly solve this cleanly. It _could_ serve as a work around to this, but it's utility is for other reasons. The core of this Github issue would be solved by using a string type. 

(The "label - value pair" suggestion could be useful for users who want to 1.) store a non-display value, such as a value that will work cleanly in a URL, formatted as CONSTANT_STYLE_VALS, etc, and 2.) another that is human friendly. It offers different functionality. Using this to solve the issue in this Github thread would be a bit hacky IMO because it would add maintenance overhead for users to have to maintain a mapping of values in their client frontend code too...unless that was their intention in the first place, which is what a "label - value pair" would provide for users.) TDLR: that's tangential, it's potentially useful, but this Github issue is better solved by storing as a string. 

Thanks for re-opening.

**Comment by derrickmehaffy** [comment]
> it would add maintenance overhead for users to have to maintain a mapping of values in their client frontend code too...unless that was their intention in the first place,

Typically that is how enumeration values are handled they are usually statically assigned mapping values. Basically the intention as I understand it from how alex put it would be to handle enums like the models are currently being handled where the display name differs from the actual model name.

`Blog Post` becomes `blog-post` in the backend for models with the reference label being saved into the `core_store` (I'd need to double check where that is stored).

---

**Edit:** To add to this you can't currently change the enumeration values in production or staging as is since those values are stored in the model settings. I'd imagine the flow for modifying these values would follow a similar migration and deployment plan for the frontend as you would in the backend with Strapi.

For string based values this is where the oneWay/manyWay relations come in, although I do understand a bit of the frustration with POST/PUT requests and these two relations meaning you need to know the ID, benefit being though on your frontend you can easily grab these values (or statically map them) while also retaining the ability to update them in production as needed.

A good example I can think of related to enumeration values mapping from frontend to backend is ServiceNow which is probably the top intranet ticket tracking system. I know the ticket status enumeration values used in that application employ a similar system

**Comment by websocket98765** [comment]
And that's one way to attempt to address it, but it adds complexity and potential bugs. It's just cleaner solved to store using the data type that it really is, a string.

**Comment by derrickmehaffy** [comment]
It is a string in essence you are right, although there has been discussion on whether or not the SQL database mappings between the field type and column type in the database should be moved over from varchar to the actual enumeration type. I think it really comes down to thinking ahead when validations come around and trying to play nice with other standards (like GraphQL)

**Comment by websocket98765** [comment]
> whether or not the SQL database mappings between the field type and column type in the database should be moved over from varchar to the actual enumeration type

That would be a bad idea from a DB-compatibility perspective and would serve only to increase complexity and bug proneness. [Enums don't even exist as a data type in MongoDB](https://www.javatpoint.com/mongodb-data-types). Mongo just stores them as strings. 

It's just a dropdown of text values, let's not get so caught up on the word `enum` that we end up with the wrong solution.

**Comment by derrickmehaffy** [comment]
The problem being though Strapi works not just with MongoDB but also MySQL, PostgreSQL, SQLite, and probably MS-SQL in the near future. So all possible databases have to be considered for compatibility.

enumeration is a programming term 🤔

**Comment by websocket98765** [comment]
> The problem being though Strapi works not just with MongoDB but also MySQL, PostgreSQL, SQLite, and probably MS-SQL in the near future. So all possible databases have to be considered for compatibility.

Precisely my point. They all support strings in one form or another. They don't all support enum data types.

**Comment by derrickmehaffy** [comment]
Well based on mongoose there is actually support for enum's (mind you in a similar way to how SQL itself checks validation): https://mongoosejs.com/docs/validation.html#built-in-validators

**Comment by derrickmehaffy** [comment]
It would also seem that MongoDB itself also has validation for enumeration values (albeit they do allow spaces where SQL does not AFAIK): https://docs.mongodb.com/manual/core/schema-validation/#json-schema

```
...
            major: {
               enum: [ "Math", "English", "Computer Science", "History", null ],
               description: "can only be one of the enum values and is required"
            },
...
```

**Comment by websocket98765** [comment]
And if you look closely at your link, it's a what? A string. Because MongoDB has strings, not enums.
```js
  drink: {
    type: String,
    enum: ['Coffee', 'Tea'],
    required: function() {
      return this.bacon > 3;
    }
  }
```

There doesn't seem to be a point to this and frankly I'm busy. I think I've clearly presented my reasoning. Happy to respond to Strapi devs after they've have time to consider it.

**Comment by derrickmehaffy** [comment]
Looking at this it would seem we could remap the values within graphql, that may provide some better compatibility: https://medium.com/atheros/how-to-design-graphql-mutations-and-queries-part-2-enums-ebb01613832

```
import {
  GraphQLEnumType,
} from 'graphql';

const taskStateEnumType = new GraphQLEnumType({
  name: 'TaskStateEnum',
  values: {
    ASSIGNED: {
      value: 0,
    },
    UNASSIGNED: {
      value: 1,
    },
    IN_PROGRESS: {
      value: 2,
    },
  },
});

export {
  taskStateEnumType as default,
};
```

Given in this case they are mapping them to integer values which is how SQL typically stores enum values (mongo itself can also do this)

**Comment by websocket98765** [comment]
That is so overly complex for a dropdown of predefined string values. Like it's not that hard.

**Comment by derrickmehaffy** [comment]
I understand where you are coming from @websocket98765 I really do but not everyone looks at it this way. I am just trying to brain storm other solutions, in the past the Strapi team have effectively said "we are following graphql". I'm not trying to be combative, I am simply trying to spark constructive discussion since this issue has been marked as a discussion topic.

I'm offering alternative ideas to try and match up with as many possible solutions to the problem as possible. Yes you are right it's an array of strings, but we need to consider compatibility with everyone not just the select few.

**Comment by alexandrebodin** [comment]
@derrickmehaffy nice find for the graphql enum mapping. This is the thing that will fix the issue. 

@websocket98765 enum are already stored as string. That is not the issue here. Not allowing certain characters was to support graphql enum types without having to handle it differently than rest. 

What @derrickmehaffy  found does solve the problem. We will be able to solve this pretty easily. allowing basically any string in the enum.

On the rest side the value and select label will be the value given by the user. With GraphQL we will convert it to a CAPTILAIZED_CLEAN const and use the mapping to convert it to the intended value. 

---

IMO using special chars in an enum or a string that is considered to be a constant  isn't a best practice. 

Dumb example but still:
```js
const HEWLETT_PACKARD= 'Hewlett Packard';
const OTHER= 'Some random @&paz,dpai,zd'

if (enum === OTHER) {
// do sth
}
```

We willl still allow it but adding a  label - value pair type will make a lot of sense for people wanting to display a nice value to there content editors and still get a clean normalized value in there API and there code.

We might make the value optional for people who want more power for example.

**Comment by derrickmehaffy** [comment]
The question then @alexandrebodin is what are your thoughts on saving the value to the database? Saving the user response but returning the `CAPTILAIZED_CLEAN` or will GQL users just have to use the `CAPTILAIZED_CLEAN` all the time? (Guess we will have to play with this, personally I'd be fine if we just stored `CAPTILAIZED_CLEAN` to the database though there would need to be a migration process built in for existing users. And by storing I mean both REST and GQL storing the `CAPTILAIZED_CLEAN` 🤔 )

**Comment by alexandrebodin** [comment]
I think that we should save the user value just for now. This avoids too much processing while still enabling people who want clean values without adding a constraint for them. (e.g capitalizing when the value is already clean). This will also make the migration process easier.

For GraphQL users there are no other choice but to use the CAPITALIZED_CLEAN value in there queries and mutations. But I would save the user input just for consistency between the content manager value and the REST api.

**Comment by derrickmehaffy** [comment]
Makes sense, I agree. We will have to play with it to see and make sure it works the way we think it will.

**Comment by websocket98765** [comment]
I'll look forward to trying it out, though I don't relish the prospect of maintaining mappings of the values back into human-friendly formats everywhere used (not DRY; e.g. in Strapi's model definition, user front end where API is consumed, all scripts that select from the DB such as mine to convert DB data to send to the search DB which must contain human friendly strings for search purposes, etc). 

Almost feels like a simple 'dropdown of strings' and enum should be different--the first for content (like brand names) and the second for constant / state values (e.g. DRAFT, PUBLISHED, etc).

**Comment by StartMeUp** [comment]
Hi guys
I ran into the issue while trying to enter values of more than one word, and as this is a discussion thread I thought I might add my point of view.
Disclaimer: I'm a real newby developer, really new in node, api and all the stuff

After reading the whole thread, with lots of technicalities, a very simple thought came to me.

As a developper, would you develop for your app a feature that is not usable ? I mean, what's the point of this enum field stuff builder if you can't enter the values you need ?

Sorry for being that basic, but I'm still amazed by so much technical discussion about something that is meant to be not usable in the end.

**Comment by lauriejim** [comment]
Hello @StartMeUpLille !
To avoid any database errors (SQL and NoSQL databases) we are not allowing space, special characters, etc..
After that, the recommended way to use the enum type is to math a value to your list to a translation key that will display something to your front end application.

**Comment by derrickmehaffy** [comment]
@StartMeUpLille Enumeration values throughout programming have generally been used for simplistic programmatic usage, not generally something used without mapping in the frontend.

In most cases you would be better served using either the `oneWay` or `manyWay` relation type. The reason for statically mapping outside of Strapi is you can't change enumeration values in production, they are hard coded into the model settings file, using relations allows you to add new values while your application is running.

A good example of using an enum would be say a gender selector:

- male
- female
- other

Something that is very strict and very unlikely to change, the current restrictions come from not only GraphQL but also the databases themselves. At the moment yes SQL databases are using strings (I was the one who suggested that, instead of using the actual enumeration type in MySQL due to the lack of migrations in Strapi currently).

As mentioned in the OP as well as https://github.com/strapi/strapi/issues/4780#issuecomment-571259480 there is a work around for the GraphQL stuff and mapping values however this issue isn't deemed a priority one from what I understand (as it's not breaking current functionality).

**Comment by websocket98765** [comment]
This has unfortunately been such a pain, and maintaining mappings of all values is so much overhead, that I'm switching any enums that are content (rather than merely a few statuses) to text inputs and using a cron job to run validation on them. It's such a hack when merely wanting a dropdown of pre-defined text values, but at least this way I only need to maintain the list of valid strings in _one_ location in my code, it's way more time efficient than dealing with creating & maintaining mappings of values everywhere when I just want a predefined string. 
: ( 

Total pain, but tired of arguing it and have to keep my project moving forward.

**Comment by Jamsie100** [comment]
It sucks that we are limited, but I found a work around to keep everything on Strapi. I create a new content-type with a json field in it. In the json field i specify the mapping I need. And on the front end application I fetch both the json config mapping and the original contenttype fields and have it map the ones specified.

example JSON

```
{
  "enumMapping": {
    "listingType": {
      "one": "1 bedroom",
      "two": "2 bedroom",
      "three": "3 bedroom",
      "shared": "private room (shared place)",
      "studio": "studio",
      "fourplus": "4+ bedrooms"
    },
}
```

example enum values

studio
one
two
three
fourplus
shared

**Comment by derrickmehaffy** [comment]
@FeiShepherd awesome workaround for now!

**Comment by harmenjanssen** [comment]
For what it's worth: my team is coming fresh to Strapi and encountering the same problems. Since we use Dutch terms in our CMS and want to include spaces and diacritical marks, I cannot use the Enumeration type. 
We would really be helped by a decoupling of the widget and the database — frankly it shouldn't matter what form widget is used to generate a database value, I'd say.
I just want to help admins prevent typos, that's all.

**Comment by derrickmehaffy** [comment]
This may end up needing some mapping system same as the CTM does current, having some kind of map from `This value` to `thisValue` (Admin Panel view vs logical backend view). The core issue is still GraphQL limitations that Strapi doesn't have much control over. Enumerations are still very strict in many different languages (not just GQL) so mapping will still be needed in the frontend as well.

In the short term @harmenjanssen I still recommend you use oneWay relation instead of enumeration.

**Comment by harmenjanssen** [comment]
I wouldn't implement a map honestly, but just allow authors to specify key/value pairs. Then you don't need inflection magic (which sooner or later will prove incorrect for certain cases, see also the incorrect pluralisations) and authors have full control over the values stored in the database.

**Comment by Aurelsicoko** [comment]
What are your thoughts about something like that?

![Screenshot 2020-04-09 at 15 42 32](https://user-images.githubusercontent.com/4144726/78901836-33fbe880-7a79-11ea-8859-d92cda9b2cb2.png)

**Comment by derrickmehaffy** [comment]
> What are your thoughts about something like that?
> 
> ![Screenshot 2020-04-09 at 15 42 32](https://user-images.githubusercontent.com/4144726/78901836-33fbe880-7a79-11ea-8859-d92cda9b2cb2.png)

@Aurelsicoko are you thinking of storing the label in the core_store or the model settings?

**Comment by soupette** [comment]
> What are your thoughts about something like that?
> 
> ![Screenshot 2020-04-09 at 15 42 32](https://user-images.githubusercontent.com/4144726/78901836-33fbe880-7a79-11ea-8859-d92cda9b2cb2.png)

I like the idea of having a new UI for creating the enum's values however I think that this component will require a lot of scrolling in the modal if we set for instance like 10 values... I think it would be great to add a way for the users the visualise all the values already created

**Comment by harmenjanssen** [comment]
> What are your thoughts about something like that?
> 
> ![Screenshot 2020-04-09 at 15 42 32](https://user-images.githubusercontent.com/4144726/78901836-33fbe880-7a79-11ea-8859-d92cda9b2cb2.png)

This UI would be great for our purposes! 

> @Aurelsicoko are you thinking of storing the label in the core_store or the model settings?

And for what it's worth: I personally would really like as much as possible to be stored in the model settings, so that I can prepare the complete CMS UI with descriptions, labels and view settings locally, and then push it out to production and have it be perfect without having to follow my deploy with a login to production and manual tweaking while the client waits.

**Comment by websocket98765** [comment]
That looks nice. I assume `Label` is the display text and `Value` the no-space allowed version that's stored in the DB. If so, maybe it'd be preferable to put `Value` on the left and `Label` on right. This would provide mental consistency with NodeJS objects, in which keys are not allowed to have spaces (unless quoted) and their string values may have spaces.

> And for what it's worth: I personally would really like as much as possible to be stored in the model settings, so that I can prepare the complete CMS UI with descriptions, labels and view settings locally, and then push it out to production and have it be perfect without having to follow my deploy with a login to production and manual tweaking while the client waits.

+1  In my usage, I prefer to define everything in the model settings JSON, so probably wouldn't actually use the UI myself.

**Comment by derrickmehaffy** [comment]
> That looks nice. I assume `Label` is the display text and `Value` the no-space allowed version that's stored in the DB. If so, maybe it'd be preferable to put `Value` on the left and `Label` on right. This would provide mental consistency with NodeJS objects, in which keys are not allowed to have spaces (unless quoted) and their string values may have spaces.
> 
> > And for what it's worth: I personally would really like as much as possible to be stored in the model settings, so that I can prepare the complete CMS UI with descriptions, labels and view settings locally, and then push it out to production and have it be perfect without having to follow my deploy with a login to production and manual tweaking while the client waits.
> 
> +1  In my usage, I prefer to define everything in the model settings JSON, so probably wouldn't actually use the UI myself.
> 

I agree on both counts here, values as the key, label as the "value" and storing this in the model settings (since you can't add/remove enums in other environments anyway)

**Comment by judehunter** [comment]
I encountered this problem as well.
What about an another `<select>` type of field but with an underlying string? I understand there is some enum logic here, but in my (and I assume many others') use case, I only need an employee to be able to select from one of many values and store that as a string.

**Comment by derrickmehaffy** [comment]
> I encountered this problem as well.
> What about an another `<select>` type of field but with an underlying string? I understand there is some enum logic here, but in my (and I assume many others') use case, I only need an employee to be able to select from one of many values and store that as a string.

@judehunter as I suggested earlier in this thread, for now you should use another model and a oneWay relation. This will work as you are needing.

**Comment by bernhardw** [comment]
I think we should allow locales (`/config/locales/xx_xx.json`) to "translate" enum values. Because "Label" in the screens above need to change depending on the current admins language.

**Reply by derrickmehaffy** [reply]
That is going to complicate things quite a bit :thinking:

**Comment by juanlanus** [comment]
Also, the text of the associated "error" message is what a programmer would tell her boss, not what a UI should tell a user "regex doesn't allow spaces".
For a normal person (a non-IT human, of which there are many) the regex concept is an annoying unknown.

**Reply by laurenskling** [reply]
I got this message: "The value does not match the regex." with no idea what the regex was, so no idea what char was actually failing

**Reply by juanlanus** [reply]
@laurenskling It's a developer-devised message, focused on the code instead of the user, as usability requires. 
For example, this message is unusable for most of the humans who have no idea of the reaming of "regex". 
It is also barely usable for those who know about regexes, as the atual one is not shown.

**Reply by laurenskling** [reply]
that's my point. What regex? I have no idea what regex is on the field.

**Reply by derrickmehaffy** [reply]
https://github.com/strapi/strapi/discussions/4780#discussioncomment-5044

**Reply by laurenskling** [reply]
Yeah I know _now_ 😅 I'd like to have learned it easier than a github comment ;) For example it's not in the docs anywhere.

**Reply by derrickmehaffy** [reply]
Trust me it's on our list of stuff for the doc rewrite :stuck_out_tongue:

**Reply by juanlanus** [reply]
@derrickmehaffy Fixing the spaces restriction is one thing.  
But fixing the redaction of the error message is another one, quick and safe to do.  
Instead of the current msg it could read something like "**entries here can only contain letters, numbers and underscores, and must start with a letter**" which is what the /[_A-Za-z][_0-9A-Za-z]*/ regex enforces - but in a human language  :-)

**Reply by derrickmehaffy** [reply]
@juanlanus can you open an issue/pull request. Looking at the source code there is a couple of references and we should probably poke at the frontend team to have them clarify which one we need to change.

`packages/strapi-plugin-content-manager/admin/src/translations/en.json` L113
`packages/strapi-helper-plugin/lib/src/testUtils/commonTrads.json` L119
`packages/strapi-admin/admin/src/translations/en.json` L287

**Reply by juanlanus** [reply]
@derrickmehaffy Sorry, I'm not entitled to do this, I'm a usability guy, and never posted a PR. 
It's not so hard, you might have there a developer who works full time with this codebase and is willing to help the users like @laurenskling to understand this error.

**Comment by websocket98765** [comment]
I switched off Strapi. But the two alternatives I evaluated, and one I chose, both store dropdowns/"single selects" as strings, like my [original suggestion](https://github.com/strapi/strapi/discussions/4780#discussioncomment-5047). Easy that way. All the best with this.

**Reply by eliasko** [reply]
What was your two alternatives you evaluated?

**Comment by derrickmehaffy** [comment]
Hi all, I'm going to close/lock this discussion in favor of our global tracking issue related to enums in general.

See: https://github.com/strapi/strapi/issues/7904
